### ISSUE TITLE
Different approach with a few benefits

### ISSUE THREAD

**Issue by notEvil** [pyutils/line_profiler#230]
Hi,

I've been tinkering with line_profiler and implemented a different approach to line profiling:

https://github.com/notEvil/line_profiler/tree/sub

Essentially, it keeps track of "calls" (`PyTrace_CALL` and `PyTrace_RETURN` are used for generators and async as well) and by doing so records "total time": time spent within a function call, generator or async step, excluding time spent in line profiled functions that were called. From there it calculates "cumulative time".

This approach has multiple benefits:
- "total time"
- times in recursive function calls are attributed correctly (see comparisons below)
- times add up (see https://github.com/pyutils/line_profiler#frequently-asked-questions)
- calls can be traced
  - the prototype creates simple statistics

Potential downside is increased overhead, but in my testing it wasn't much slower.

I've also created a Neovim plugin (not published yet) to see the differences:

<details>

```python
import time


@profile
def f():
    time.sleep(0.1)
    (g(1), g(2))
    g(3)


@profile
def g(n):
    if n == 0:
        return

    time.sleep(0.1)
    g(n - 1)


f()
```

Original
![Screenshot from 2023-08-12 13-35-01 crop](https://github.com/pyutils/line_profiler/assets/4567628/d554a546-4d45-4050-9066-113a783be015)
New cumulative
![Screenshot from 2023-08-12 13-35-21 crop](https://github.com/pyutils/line_profiler/assets/4567628/ea3c8b60-7283-4b86-9e36-43d22478d1e4)
New total
![Screenshot from 2023-08-12 13-35-37 crop](https://github.com/pyutils/line_profiler/assets/4567628/86c6c160-88a5-4e77-aecd-9c6a877bbc3c)

</details>

Tested on a real application (40s runtime, no async)

**Comment by Erotemic** [comment]
It would help to see the `python -m line_profiler <script>.lprof` results in addition to the neovim (heatmap?) visualization to get a better sense of what is happening and what the difference is.

**Comment by notEvil** [comment]
There you go

<details>

Original
```
Timer unit: 1e-06 s

Total time: 0.70161 s
File: /home/arappold/git/line_profiler/sub.py
Function: f at line 4

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     4                                           @profile
     5                                           def f():
     6         1     100060.9 100060.9     14.3      time.sleep(0.1)
     7         1     300288.5 300288.5     42.8      (g(1), g(2))
     8         1     301260.9 301260.9     42.9      g(3)

Total time: 0.601441 s
File: /home/arappold/git/line_profiler/sub.py
Function: g at line 11

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    11                                           @profile
    12                                           def g(n):
    13         9          6.2      0.7      0.0      if n == 0:
    14         3          0.9      0.3      0.0          return
    15                                           
    16         6     601357.4 100226.2    100.0      time.sleep(0.1)
    17         6         76.5     12.8      0.0      g(n - 1)

```
New cumulative
```
Timer unit: 1e-06 s

Total time: 0.701555 s
File: /home/arappold/git/line_profiler/sub.py
Function: f at line 4

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     4                                           @profile
     5                                           def f():
     6         1     100061.0 100061.0     14.3      time.sleep(0.1)
     7         1     300265.7 300265.7     42.8      (g(1), g(2))
     8         1     301228.6 301228.6     42.9      g(3)

Total time: 1.00325 s
File: /home/arappold/git/line_profiler/sub.py
Function: g at line 11

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    11                                           @profile
    12                                           def g(n):
    13         1          7.2      7.2      0.0      if n == 0:
    14         1          3.6      3.6      0.0          return
    15                                           
    16         1     601360.9 601360.9     59.9      time.sleep(0.1)
    17         1     401881.4 401881.4     40.1      g(n - 1)

```
The prototype doesn't yet count the number of hits (not difficult to add)
</details>

**Comment by Theelx** [comment]
That looks promising, but the tracing seems to have a 4x overhead based on the return function in the base case (assuming it was hit 3 times in your cumulative version). While that's not too significant in this case, it can be very significant in tighter loops, and it would be beneficial to create a tight loop to test this in.

**Comment by notEvil** [comment]
How do you infer a 4x overhead from that output? I just tested

```python
@profile
def f():
    for _ in range(int(1e7)):
        g()


@profile
def g():
    pass


f()
```

and it takes 22.37s with v4.0.3 and 18.53s with the new approach. The prototype runs the new approach in addition to the original one, so I temporarily disabled the original part for this crude benchmark.

**Comment by Theelx** [comment]
Ah, I looked at the return statement in the base case and it looked like it took 1.2us per hit in your version whereas the original took 0.3us per hit. Usually, very simple lines like a plain return statement are virtually instant in Python, and almost all the time taken is in the profiler.

**Comment by Theelx** [comment]
Here's a benchmark that shows around a 11x slowdown with the current line_profiler vs unprofiled code (total time 500ms unprofiled vs 5.4s profiled):
```py
Time Elapsed tight: 494.96ms
Time Elapsed tight: 5390.15ms
Slowdown from profiling: 4895.19ms, 10.89x

Total time: 5.39021 s
File: tight.py
Function: main at line 3

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     3                                           def main():
     4         1          3.1      3.1      0.0      start = time.perf_counter()
     5  20000001    2628245.8      0.1     48.8      for x in range(20000000):
     6  20000000    2761886.3      0.1     51.2          y = x
     7         1         31.5     31.5      0.0      elapsed_ms = round((time.perf_counter()-start)*1000, 2)
     8         1         42.2     42.2      0.0      print(f"Time Elapsed tight: {elapsed_ms}ms")
     9         1          1.3      1.3      0.0      return elapsed_ms
```
Your code gives these results:
```py
Time Elapsed tight: 494.96ms
Time Elapsed tight: 15165.31ms
Slowdown from profiling: 14670.35ms, 30.64x

Total time: 6.53584 s
File: tight.py
Function: main at line 3

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     3                                           def main():
     4         1          1.4      1.4      0.0      start = time.perf_counter()
     5  20000001    3116578.9      0.2     47.7      for x in range(20000000):
     6  20000000    3419187.5      0.2     52.3          y = x
     7         1         23.3     23.3      0.0      elapsed_ms = round((time.perf_counter()-start)*1000, 2)
     8         1         51.6     51.6      0.0      print(f"Time Elapsed tight: {elapsed_ms}ms")
     9         1          0.6      0.6      0.0      return elapsed_ms
```
I believe the total time elapsed shown by line_profiler is incorrect, as I definitely waited at least 14 seconds for yours to finish (based off watching the clock). This indicates that your code has a ~30.5x slowdown, which is ~3x more overhead than the current line_profiler.

**Comment by Theelx** [comment]
The thing to remember (which I forgot in my first comment here!) is that line_profiler's measurements of time taken in the output aren't actually the overhead, as line_profiler calls hpTimer() to try to eliminate the overhead from the output.

**Comment by notEvil** [comment]
I fail to reproduce your results. Could you try to run them again with the new commit. I changed a few things so you can get a fair comparison with your benchmark.

**Comment by Theelx** [comment]
This version is significantly better, but still about 1.5x slower:
```py
Time Elapsed tight: 513.24ms
Time Elapsed tight: 9631.74ms
Slowdown from profiling: 9118.5ms, 18.77x
Wrote profile results to tight.py.lprof
Timer unit: 1e-06 s

Total time: 4.17773 s
File: /home/theel/Coding/line_profiler/line_profiler_cumulative/tight.py
Function: main at line 3

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     3                                           def main():
     4         1          1.7      1.7      0.0      start = time.perf_counter()
     5         1    2005221.9 2005221.9     48.0      for x in range(20000000):
     6         1    2172444.6 2172444.6     52.0          y = x
     7         1         21.0     21.0      0.0      elapsed_ms = round((time.perf_counter()-start)*1000, 2)
     8         1         38.0     38.0      0.0      print(f"Time Elapsed tight: {elapsed_ms}ms")
     9         1          0.7      0.7      0.0      return elapsed_ms
```

**Comment by notEvil** [comment]
The weird thing is, my results are the complete opposite. v4.0.3 gives me "Time Elapsed tight: 12030.22ms" while the new commit "Time Elapsed tight: 6169.69ms" (tripple checked). My process
- checkout tag/branch
- `pipenv --rm`
- `pipenv run pip install /path/to/repo` (Python 3.11.3)
- `pipenv run kernprof -l ...`

**Comment by Theelx** [comment]
Ohhhhh, the issue is that I was using my fork from #203 instead of the current one. I can confirm that yours is faster than the version of line_profiler in this repo, but it would be slower if #203 were to be merged.

**Comment by notEvil** [comment]
Great to hear :) I think most improvements can be carried over. I'll give it a try sometime next week

**Comment by Theelx** [comment]
See https://github.com/notEvil/line_profiler/pull/1

**Comment by notEvil** [comment]
Nice, thanks! So if I understand correctly, performance is fine.

Regarding the differences to the current approach, the implementation at the bottom of [`_line_profiler.pyx`](https://github.com/pyutils/line_profiler/compare/main...notEvil:line_profiler:sub#diff-06943e1df2f9908c2330ae5f03985e808db297dbe0440f36c1f720e6474dcb7f) is supposed to be readable. Just ignore the parts about `call`. If its not clear, I can add comments pointing out its details.

**Comment by notEvil** [comment]
The latest commit now counts the number of line hits and provides the times in a backwards compatible way.

```
Timer unit: 1e-06 s

Total time: 0.700474 s
File: /home/arappold/git/line_profiler/sub.py
Function: f at line 4

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     4                                           @profile
     5                                           def f():
     6         1     100057.6 100057.6     14.3      time.sleep(0.1)
     7         1     300195.1 300195.1     42.9      (g(1), g(2))
     8         1     300221.3 300221.3     42.9      g(3)

Total time: 1.00071 s
File: /home/arappold/git/line_profiler/sub.py
Function: g at line 11

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    11                                           @profile
    12                                           def g(n):
    13         9          2.4      0.3      0.0      if n == 0:
    14         3          0.5      0.2      0.0          return
    15                                           
    16         6     600365.7 100060.9     60.0      time.sleep(0.1)
    17         6     400342.9  66723.8     40.0      g(n - 1)
```

"Total time" reported for `g` is wrong because its summing up cumulative times which don't add up for recursive functions. Total times would, but the default output should remain cumulative times. So maybe remove the "% Time" column if its a recursive function (can be inferred from call stats)?

**Comment by notEvil** [comment]
An example of what call stats would allow: [graph.dot.pdf](https://github.com/pyutils/line_profiler/files/12329317/graph.dot.pdf)
created with https://github.com/notEvil/line_profiler/blob/sub/graph.py

**Comment by notEvil** [comment]
I'm happy to announce the next iteration :) Getting useful numbers in case of recursive code was a lot more difficult than anticipated, but here they are nonetheless. The last commit
- adds `block_times`: how much time is spent in each function,
- adds `call_times`: how much time is spent in the function called
- and fixes "cumulative time" of lines

It is impossible to derive `block_times` and `call_times` from the line timings nor from call stats since in the presence of recursion, cumulative times don't add up.

Now the output for the initial example is

```
Timer unit: 1e-06 s

Total time: 0.700534 s
File: /home/arappold/git/line_profiler/example1.py
Function: f at line 4

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     4                                           @profile
     5                                           def f():
     6         1     100061.6 100061.6     14.3      time.sleep(0.1)
     7         1     300238.6 300238.6     42.9      (g(1), g(2))
     8         1     300233.5 300233.5     42.9      g(3)

Total time: 0.600456 s
File: /home/arappold/git/line_profiler/example1.py
Function: g at line 11

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    11                                           @profile
    12                                           def g(n):
    13         9          3.6      0.4      0.0      if n == 0:
    14         3          0.8      0.3      0.0          return
    15                                           
    16         6     600380.3 100063.4    100.0      time.sleep(0.1)
    17         6     300267.2  50044.5     50.0      g(n - 1)
```

<details>
<summary>Testet with</summary>

<details>
<summary>Code</summary>

```python
import time


def wait(n):
    time.sleep(n * 0.01)


@profile
def f(n):
    f1(n)
    f2(n)
    f3(n)
    f4(n)
    f5(n)
    f6(n)


@profile
def f1(n):
    if n == 0:
        return

    wait(1)
    f1(n - 1)


@profile
def f2(n):
    if n == 0:
        return

    wait(1)
    (f2(n - 1), f2(n // 2))


@profile
def f3(n):
    if n == 0:
        return

    wait(1)
    f3_2(n - 1)


@profile
def f3_2(n):
    if n == 0:
        return

    wait(1)
    f3(n - 1)


@profile
def f4(n):
    if n == 0:
        return

    wait(1)
    (f4_2(n - 1), f4_2(n // 2))


@profile
def f4_2(n):
    if n == 0:
        return

    wait(1)
    (f4(n - 1), f4(n // 2))


@profile
def f5(n):
    if n == 0:
        return

    wait(1)
    f5_2(n - 1)


@profile
def f5_2(n):
    if n == 0:
        return

    wait(1)
    f5_3(n - 1)


@profile
def f5_3(n):
    if n == 0:
        return

    wait(1)
    f5(n - 1)


@profile
def f6(n):
    if n == 0:
        return

    wait(1)
    (f6_2(n - 1), f6_2(n // 2))


@profile
def f6_2(n):
    if n == 0:
        return

    wait(1)
    (f6_3(n - 1), f6_3(n // 2))


@profile
def f6_3(n):
    if n == 0:
        return

    wait(1)
    (f6(n - 1), f6(n // 2))


f(6)
```
</details>

[graph.dot.pdf](https://github.com/pyutils/line_profiler/files/12361167/graph.dot.pdf)

<details>
<summary>Output</summary>

```
Timer unit: 1e-06 s

Total time: 0.754708 s
File: /home/arappold/git/line_profiler/sub.py
Function: f at line 8

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     8                                           @profile
     9                                           def f(n):
    10         1      60377.5  60377.5      8.0      f1(n)
    11         1     191180.4 191180.4     25.3      f2(n)
    12         1      60361.0  60361.0      8.0      f3(n)
    13         1     191203.2 191203.2     25.3      f4(n)
    14         1      60369.8  60369.8      8.0      f5(n)
    15         1     191216.3 191216.3     25.3      f6(n)

Total time: 0.0603742 s
File: /home/arappold/git/line_profiler/sub.py
Function: f1 at line 18

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    18                                           @profile
    19                                           def f1(n):
    20         7          1.3      0.2      0.0      if n == 0:
    21         1          0.1      0.1      0.0          return
    22                                           
    23         6      60350.0  10058.3    100.0      wait(1)
    24         6      50312.4   8385.4     83.3      f1(n - 1)

Total time: 0.191177 s
File: /home/arappold/git/line_profiler/sub.py
Function: f2 at line 27

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    27                                           @profile
    28                                           def f2(n):
    29        39          4.8      0.1      0.0      if n == 0:
    30        20          1.1      0.1      0.0          return
    31                                           
    32        19     191067.2  10056.2     99.9      wait(1)
    33        19     181120.1   9532.6     94.7      (f2(n - 1), f2(n // 2))

Total time: 0.0603579 s
File: /home/arappold/git/line_profiler/sub.py
Function: f3 at line 36

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    36                                           @profile
    37                                           def f3(n):
    38         4          0.6      0.1      0.0      if n == 0:
    39         1          0.1      0.1      0.0          return
    40                                           
    41         3      30169.0  10056.3     50.0      wait(1)
    42         3      50301.2  16767.1     83.3      f3_2(n - 1)

Total time: 0.0502975 s
File: /home/arappold/git/line_profiler/sub.py
Function: f3_2 at line 45

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    45                                           @profile
    46                                           def f3_2(n):
    47         3          0.4      0.1      0.0      if n == 0:
    48                                                   return
    49                                           
    50         3      30168.7  10056.2     60.0      wait(1)
    51         3      40241.4  13413.8     80.0      f3(n - 1)

Total time: 0.191199 s
File: /home/arappold/git/line_profiler/sub.py
Function: f4 at line 54

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    54                                           @profile
    55                                           def f4(n):
    56        21          2.8      0.1      0.0      if n == 0:
    57        12          0.7      0.1      0.0          return
    58                                           
    59         9      90512.6  10057.0     47.3      wait(1)
    60         9     181143.0  20127.0     94.7      (f4_2(n - 1), f4_2(n // 2))

Total time: 0.181137 s
File: /home/arappold/git/line_profiler/sub.py
Function: f4_2 at line 63

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    63                                           @profile
    64                                           def f4_2(n):
    65        18          2.2      0.1      0.0      if n == 0:
    66         8          0.4      0.0      0.0          return
    67                                           
    68        10     100576.7  10057.7     55.5      wait(1)
    69        10     161017.3  16101.7     88.9      (f4(n - 1), f4(n // 2))

Total time: 0.060366 s
File: /home/arappold/git/line_profiler/sub.py
Function: f5 at line 72

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    72                                           @profile
    73                                           def f5(n):
    74         3          0.4      0.1      0.0      if n == 0:
    75         1          0.1      0.1      0.0          return
    76                                           
    77         2      20113.6  10056.8     33.3      wait(1)
    78         2      50308.6  25154.3     83.3      f5_2(n - 1)

Total time: 0.0503043 s
File: /home/arappold/git/line_profiler/sub.py
Function: f5_2 at line 81

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    81                                           @profile
    82                                           def f5_2(n):
    83         2          0.3      0.2      0.0      if n == 0:
    84                                                   return
    85                                           
    86         2      20115.2  10057.6     40.0      wait(1)
    87         2      40247.2  20123.6     80.0      f5_3(n - 1)

Total time: 0.0402436 s
File: /home/arappold/git/line_profiler/sub.py
Function: f5_3 at line 90

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    90                                           @profile
    91                                           def f5_3(n):
    92         2          0.3      0.1      0.0      if n == 0:
    93                                                   return
    94                                           
    95         2      20113.8  10056.9     50.0      wait(1)
    96         2      30186.7  15093.3     75.0      f5(n - 1)

Total time: 0.191212 s
File: /home/arappold/git/line_profiler/sub.py
Function: f6 at line 99

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    99                                           @profile
   100                                           def f6(n):
   101        13          1.8      0.1      0.0      if n == 0:
   102         6          0.3      0.0      0.0          return
   103                                           
   104         7      70401.5  10057.4     36.8      wait(1)
   105         7     181155.2  25879.3     94.7      (f6_2(n - 1), f6_2(n // 2))

Total time: 0.181149 s
File: /home/arappold/git/line_profiler/sub.py
Function: f6_2 at line 108

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   108                                           @profile
   109                                           def f6_2(n):
   110        14          1.7      0.1      0.0      if n == 0:
   111         8          0.5      0.1      0.0          return
   112                                           
   113         6      60347.1  10057.8     33.3      wait(1)
   114         6     161035.2  26839.2     88.9      (f6_3(n - 1), f6_3(n // 2))

Total time: 0.161023 s
File: /home/arappold/git/line_profiler/sub.py
Function: f6_3 at line 117

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   117                                           @profile
   118                                           def f6_3(n):
   119        12          1.6      0.1      0.0      if n == 0:
   120         6          0.3      0.1      0.0          return
   121                                           
   122         6      60343.6  10057.3     37.5      wait(1)
   123         6     120793.6  20132.3     75.0      (f6(n - 1), f6(n // 2))
```
</details>
</details>

**Comment by Theelx** [comment]
Thanks for this great work! I'm going to rebase my PR to your fork on this latest work, could you take a look at it once I do that? After this work of yours is merged, I can start working on adding CPython 3.12 support to line_profiler.

**Comment by notEvil** [comment]
Thanks :)
I would still consider it work-in-progress. The call stats feature is somewhat out of scope of line profiling, naming of stuff breaks with former conventions (which is different from the more common convention of "total" vs "cumulative"), some parts are pretty ugly, the default output from CLI could make better use of the available information, ...
And there is performance, which has to be worse than before (when both are equally optimized).
It would be great if someone more experienced with the existing code base could review the fork and propose changes.

**Comment by Theelx** [comment]
I just tested the performance, and it is indeed significantly worse than before. Previously, https://github.com/notEvil/line_profiler/pull/1 was able to pull it down to 13x overhead over no profiling, in the tight loop case from above, whereas now with the same optimizations applied, it'd 16x overhead over no profiling. That being said, I believe I can optimize it further, as some of your data structure choices may be suboptimal performance-wise.

**Comment by tmm1** [comment]
@notEvil thanks for sharing this! many years ago i wrote a line profiler for another dynamic language runtime and I think I used a similar call/return stacking approach.

if you get a chance, would you mind taking a look at the merge conflicts pulling in the latest `main` after the recent autoprofiler changes. would love to try the two together.

**Comment by notEvil** [comment]
@tmm1 sry for not responding! I will check the conflicts soon, I promise ^^

I've pushed a new commit\* which should improve confidence in the solution. The previous version was manually tested on a few examples with manual inspection of the output. `graph.py` was very helpful, but obviously we'd want something automated with a lot more variation. So I started the journey of implementing strategies for generating code with [*Hypothesis*](https://hypothesis.readthedocs.io/en/latest/)\*\*, implementing a reference profiler and using it to validate the numbers. And it took a while to work as I wanted it to (statements spanning multiple lines, recursions, generators, async, ...), to look reasonably good and eventually for the tests to pass. So my apologies for keeping this private and not providing any updates.

The testing is not perfect:
- It might generate lots of useless examples
- It uses busy loops to simulate work
    - context switches introduce inaccuracies
    - busy loops are more accurate than `time.sleep`
    - inaccurate timings can and will eventually produce false positives
- It doesn't validate call profiles (hit counts and cumulative time) and call stats
    - a "call" (CALL RETURN pair) might be a function call or a generator/coroutine continuation
    - it is not easy to add generator/coroutine "call"s to the reference profiler
    - I don't think it makes sense to further complicate the Cython code to distinguish them

What I like:
- The generated examples look very diverse and complex
- The reference profiler is very simple

So now that it "works" and won't get any more complex, I'd say its a good time to review and benchmark.

https://github.com/notEvil/line_profiler/tree/sub

\* changes

- added counting of hits for blocks and calls
- added counting of primitive hits for lines, blocks and calls
- added randomized tests
- improved `graph.py` to show the additional information
- renamed structs `LineTime` and `Timing` to `CLineProfile` and `LineProfile` respectively
- added structs `CBlockProfile` and `CCallProfile`, and classes `BlockProfile` and `CallProfile`
- fixed minor regressions

\*\* If you don't know *Hypothesis* yet, I'd definitely recommend checking it out!

**Comment by notEvil** [comment]
@tmm1 just merged main and except for trivial conflicts it went smooth. The randomized tests still pass :) Beware though that `line_profiler.py` isn't adjusted yet and might show unexpected things, especially aggregates.

@Theelx you created a PR back then for better performance and comparisons. I didn't merge because I didn't want to mix the two proposals and it looked like it was based off of https://github.com/pyutils/line_profiler/pull/203 but without git reference (merge). Would you consider the PR mature enough to merge with sub on a test branch?

**Comment by notEvil** [comment]
Just a quick update: I tried to measure and improve performance and got some interesting results. The benchmark is https://gitlab.com/notEvil/code_collapse applied to itself which spends most of the time in Black and the remaining time in nested generators (tree traversal) and LibCST.

| name | Python 3.11 | vs baseline | Python 3.12 | vs baseline |
|--------|--------|--------|--------|--------|
| baseline | 16.1 |  | 34.3 |
| v4.1.2 | 27.2 | 168.9% | 178.6 | 520.7% |
| sub | 28.2 | 175.2% | 167.1 | 487.2% |
| key opt | 28 | 173.9% | 161.6 | 471.1% |
| tss opt\* | 27.6 | 171.4% |  |  |
| preshed opt\* | 27.2 | 168.9% |  |  |
| map opt\* | 26.3 | 163.4% |  |  |
| time opt\* | 25.8 | 160.3% |  |  |
| all opt | 25.2 | 156.5% | 150.9 | 439.9% |

- key opt: use `address of bytecode << 24 + line number` as key (with lower bits of address as offset for some randomness)
    - not worth it imo, but the first one I tried
- tss opt: store thread id in thread local storage to save some `threading.get_ident()`
- preshed opt: use [*Preshed*](https://github.com/explosion/preshed) for entry check
- map opt: use `parallel_flat_hash_map` and `flat_hash_map` from [*parallel-hashmap*](https://github.com/greg7mdp/parallel-hashmap) instead of `unordered_map`
- time opt: check time just after entry check instead of asap
- all opt: all combined, see https://github.com/notEvil/line_profiler/tree/subopt

As you can see, I wasn't able to significantly improve performance of sub. Maybe time opt is worth the tradeoff (creates some bias), what do you think?

I will try to figure out why Python 3.12 is performing so bad compared to 3.11 on my project. Regarding the bad performance when line profiling on 3.12, can you reproduce this? e.g. jsonpickle

\* either directly from https://github.com/pyutils/line_profiler/pull/203 or inspired by

**Comment by notEvil** [comment]
Good news I guess:

<details>

| name                       |   baseline |   v4.1.2 |    sub |   subopt |
|:---------------------------|-----------:|---------:|-------:|---------:|
| raytrace                   |       1.34 |    26.01 |  29.02 |    20.58 |
| scimark_sor                |       0.55 |    26.13 |  27.74 |    20.08 |
| scimark_lu                 |       0.55 |    24.15 |  25.27 |    18.99 |
| go                         |       0.63 |    30.64 |  27.46 |    18.16 |
| generators                 |       0.88 |    18.51 |  24.18 |    16.12 |
| spectral_norm              |       0.47 |    19.47 |  20.65 |    14.97 |
| pyflate                    |       1.96 |    18.55 |  16.60 |    11.75 |
| richards                   |       0.45 |    16.67 |  16.35 |    11.56 |
| richards_super             |       0.54 |    15.32 |  15.37 |    10.58 |
| pprint_pformat             |       6.59 |    15.44 |  12.09 |     7.89 |
| pprint_safe_repr           |       3.22 |    15.29 |  11.77 |     7.52 |
| nbody                      |       0.46 |    13.91 |   9.75 |     6.64 |
| scimark_monte_carlo        |       0.62 |    11.62 |   9.59 |     6.24 |
| fannkuch                   |       1.71 |    14.26 |   9.42 |     6.16 |
| chaos                      |       0.62 |     8.48 |   7.66 |     5.53 |
| float                      |       0.74 |     8.26 |   7.71 |     5.53 |
| tomli_loads                |       9.61 |     5.97 |   7.92 |     5.47 |
| scimark_fft                |       1.45 |    10.06 |   7.04 |     4.79 |
| meteor_contest             |       0.48 |     9.76 |   6.54 |     4.33 |
| async_generators           |       1.68 |     4.93 |   5.99 |     4.20 |
| async_tree_none_tg         |       2.15 |     6.96 |   5.93 |     3.95 |
| async_tree_memoization_tg  |       2.89 |     6.27 |   5.60 |     3.74 |
| async_tree_memoization     |       3.03 |     5.93 |   5.46 |     3.67 |
| sqlglot_normalize          |       0.50 |     3.93 |   5.27 |     3.66 |
| async_tree_none            |       2.45 |     5.95 |   5.44 |     3.66 |
| async_tree_io_tg           |       5.44 |     5.89 |   5.34 |     3.54 |
| async_tree_io              |       5.61 |     5.70 |   5.20 |     3.53 |
| async_tree_cpu_io_mixed_tg |       3.26 |     5.41 |   4.89 |     3.29 |
| async_tree_cpu_io_mixed    |       3.45 |     5.18 |   4.79 |     3.13 |
| mdp                        |      12.39 |     4.11 |   4.29 |     3.01 |
| create_gc_cycles           |       1.44 |     1.98 |   2.05 |     2.94 |
| sympy_expand               |       2.19 |     2.82 |   3.68 |     2.74 |
| sympy_str                  |       1.37 |     2.73 |   3.56 |     2.68 |
| nqueens                    |       0.76 |     3.55 |   3.61 |     2.33 |
| genshi_xml                 |       0.48 |     2.83 |   3.74 |     2.32 |
| sympy_sum                  |       0.93 |     2.38 |   3.01 |     2.29 |
| tornado_http               |       0.60 |     2.45 |   2.77 |     2.24 |
| sqlalchemy_declarative     |       0.63 |     2.35 |   2.84 |     2.22 |
| docutils                   |      12.02 |     2.42 |   2.99 |     2.21 |
| xml_etree_iterparse        |       0.57 |     2.41 |   3.25 |     2.19 |
| crypto_pyaes               |       0.67 |     4.31 |   3.03 |     2.08 |
| xml_etree_process          |       0.57 |     2.51 |   3.19 |     2.07 |
| coverage                   |       0.69 |     2.02 |   2.03 |     2.01 |
| gc_traversal               |       1.20 |     2.85 |   2.27 |     1.84 |
| sqlglot_optimize           |       0.49 |     1.91 |   2.48 |     1.73 |
| html5lib                   |       0.55 |     2.24 |   2.46 |     1.70 |
| xml_etree_generate         |       0.80 |     2.04 |   2.56 |     1.69 |
| comprehensions             |       0.44 |     1.82 |   1.13 |     1.58 |
| json_dumps                 |       0.43 |     1.33 |   1.10 |     1.42 |
| sqlite_synth               |       0.59 |     1.01 |   0.90 |     1.40 |
| pidigits                   |       0.83 |     1.55 |   1.54 |     1.39 |
| xml_etree_parse            |       0.79 |     1.45 |   1.62 |     1.38 |
| telco                      |       0.51 |     1.37 |   1.00 |     1.38 |
| scimark_sparse_mat_mult    |       0.64 |     1.00 |   1.13 |     1.34 |
| typing_runtime_protocols   |       0.57 |     1.41 |   0.86 |     1.31 |
| asyncio_tcp                |       3.90 |   nan    |   1.56 |     1.30 |
| dulwich_log                |       0.67 |     1.58 |   1.68 |     1.29 |
| coroutines                 |       0.47 |     1.32 |   1.70 |     1.26 |
| deepcopy_reduce            |       0.48 |     1.16 |   0.96 |     1.18 |
| chameleon                  |       0.49 |     1.35 |   0.91 |     1.17 |
| hexiom                     |       0.47 |     1.55 |   1.60 |     1.15 |
| regex_effbot               |       0.52 |     1.27 |   1.22 |     1.13 |
| asyncio_tcp_ssl            |      11.70 |   nan    |   1.21 |     1.10 |
| 2to3                       |       1.40 |     0.98 |   0.98 |     1.07 |
| pickle                     |       0.46 |     1.19 |   1.15 |     1.07 |
| unpickle                   |       0.57 |     1.14 |   1.11 |     1.05 |
| deepcopy                   |       0.44 |     1.12 |   1.76 |     1.04 |
| unpickle_list              |       0.44 |     1.10 |   1.07 |     1.04 |
| pickle_list                |       0.72 |     0.58 |   1.08 |     1.04 |
| asyncio_websockets         |       2.60 |     1.07 |   1.03 |     1.01 |
| regex_dna                  |       0.76 |     1.00 |   1.01 |     1.01 |
| pickle_dict                |       0.63 |     1.03 |   1.02 |     1.00 |
| sqlglot_transpile          |       0.51 |     1.09 |   1.40 |     0.99 |
| mako                       |       0.78 |     0.56 |   0.71 |     0.98 |
| python_startup             |       2.33 |     0.98 |   0.98 |     0.98 |
| python_startup_no_site     |       1.88 |     0.98 |   0.97 |     0.97 |
| logging_simple             |       0.60 |     0.86 |   0.75 |     0.94 |
| logging_format             |       0.65 |     0.81 |   0.71 |     0.90 |
| logging_silent             |       0.66 |     0.78 |   0.73 |     0.85 |
| deltablue                  |       0.58 |     1.14 |   1.20 |     0.82 |
| regex_v8                   |       0.77 |     0.59 |   0.99 |     0.82 |
| pathlib                    |       0.71 |     1.10 |   1.07 |     0.74 |
| json_loads                 |       0.59 |     0.95 |   0.89 |     0.73 |
| sympy_integrate            |       0.81 |     0.75 |   0.98 |     0.72 |
| deepcopy_memo              |       0.72 |     0.81 |   0.64 |     0.72 |
| unpack_sequence            |       0.68 |     0.64 |   0.90 |     0.64 |
| genshi_text                |       0.83 |     0.69 |   1.04 |     0.63 |
| sqlalchemy_imperative      |       0.74 |     0.68 |   0.77 |     0.60 |
| sqlglot_parse              |       0.82 |     0.58 |   0.76 |     0.54 |
| bench_mp_pool              |       0.48 |   nan    | nan    |   nan    |
| bench_thread_pool          |       0.64 |   nan    | nan    |   nan    |
| dask                       |       2.03 |   nan    | nan    |   nan    |
| django_template            |       0.62 |   nan    | nan    |   nan    |
| pickle_pure_python         |       0.44 |   nan    | nan    |   nan    |
| regex_compile              |       0.61 |   nan    | nan    |   nan    |
| unpickle_pure_python       |       0.67 |   nan    | nan    |   nan    |

</details>

I managed to adjust pyperformance so it uses line_profiler (with autoprofile) and ran it\*. See https://github.com/notEvil/line_profiler/tree/sub/pyperformance and https://github.com/notEvil/pyperformance/tree/line_profiler if you want to do the same. Set up the Pipenv environment and run `prepare.py` to prepare an environment. Then run pyperformance in that environment.

The column `baseline` is the average duration in seconds, the others are scaling factors (e.g. 1.5 means 1.5 * baseline). Some benchmarks like 2to3 spawn subprocesses which explains some ~1 factors. Didn't investigate the < 1 yet.

\* on Intel 4570S with 3 of them in parallel (1 idle core). Most nan are due to conflicts (e.g. bind to port).
